src/consumer/runner.js

```diff
@@ -322,31 +323,33 @@ module.exports = class Runner extends EventEmitter {
       }
 
       enqueuedTasks.push(async () => {
-        const batches = await result.value
+        let batches
+        try {
+          batches = await result.value
```

Comment:
Not sure if my brain is just bad or if this code is awful (not your changes specifically, but rather the fetch loop in general). It's like playing 8 dimensional chess at times.

So I get the general idea is to basically try to collect any error that happens, and if there are any, abort any future batch processing before it happens. What I don't understand is that if `result.value` throws, then the processing would never get scheduled (i.e. we'd never reach `batches.map(batch => concurrently(...))`), so wouldn't this achieve the same thing?

```js
let batches
try {
  batches = await result.value
} catch (e) {
  unlockWithError(e)
}
```

If we throw the error here, the lock would never be unlocked, since the `Promise.all` would reject.

Reply:
> if result.value throws, then the processing would never get scheduled (i.e. we'd never reach batches.map(batch => concurrently(...)))

Yes, for that broker response.  But processing work has already been added to the concurrency queue from previous broker responses.  And future broker responses could also add work to the queue.

My understanding of how it works:  There are 2 things happening concurrently
1) broker fetch requests are resolving and as they do, processing tasks are added to the concurrency queue
2) concurrency queue processing
An error can occur in either of these.  In either case, we don't want to do any more processing.

You comment has made me realise that the current proposal is still not complete.  If a broker fetch fails and we throw, and the Promise.all() throws, we are still throwing/rejecting too soon - there could be some processing tasks executing (although no more will happen after those.)  We should still wait for the barrier lock to resolve.  I'll push another change to fix that.

I'll also see if a couple of comments can help here.  But if you think the code is too far gone to be readable/ maintainable, I could have a go at re-working it.